### PR TITLE
Content Modifier

### PR DESCRIPTION
**Pull request by Barackos**

### CODE REVIEW

File: packages/ricos-content/web/src/RicosContentAPI/modify.spec.ts [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+      .set(n => ({
+        ...n,
+        imageData: {
+          ...n.imageData,
+          altText: n.imageData?.altText || 'Custom alt text',
+        },
+      }));
+
+    expect(actual).toStrictEqual(expected);
+    expect(actual).not.toEqual(content);
```
Comment (lxgreen) [inline packages/ricos-content/web/src/RicosContentAPI/modify.spec.ts:30]:
```suggestion
```

Comment (lxgreen) [reply packages/ricos-content/web/src/RicosContentAPI/modify.spec.ts:30]: what's the point?

File: packages/ricos-content/web/src/RicosContentAPI/modify.spec.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
+import { cloneDeep } from 'lodash';
+import { modify } from './modify';
+import { Node_Type, RichContent } from 'ricos-schema';
+import rawContent from '../../statics/json/migratedFixtures/migration-content.json';
```
Comment (lxgreen) [inline packages/ricos-content/web/src/RicosContentAPI/modify.spec.ts:4]: i'd take a smaller content and create the modified version as well

File: packages/ricos-content/web/src/RicosContentAPI/modify.spec.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+describe('Content modify', () => {
+  it('Should add alt text for images', () => {
+    const content = RichContent.fromJSON(rawContent);
+    const expected = cloneDeep(content);
```
Comment (lxgreen) [inline packages/ricos-content/web/src/RicosContentAPI/modify.spec.ts:9]: this is considered bad practice in testing -- modifying data in the same way the tested function does

File: packages/ricos-content/web/src/RicosContentAPI/modify.spec.ts [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        },
+      }));
+
+    expect(actual).toStrictEqual(expected);
+    expect(actual).not.toEqual(content);
+  });
+
+  it('Should duplicate all dividers', () => {
+    const content = RichContent.fromJSON(rawContent);
+    const expected = cloneDeep(content);
```
Comment (lxgreen) [inline packages/ricos-content/web/src/RicosContentAPI/modify.spec.ts:35]: same here

File: packages/ricos-content/web/src/RicosContentAPI/modify.ts [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    return new TraversalModifier(
+      this.traversal.composePrism(Prism.fromPredicate(predicate)),
+      this.tree,
+      this.content
+    );
+  }
+
+  set(setter: (node: Node) => Node | Node[]) {
+    const selectedKeys = compact(this.traversal.asFold().getAll(this.tree)).map(({ key }) => key);
+    const cleverFold = T.fold<Node, Node>((value, forest) => {
```
Comment (lxgreen) [inline packages/ricos-content/web/src/RicosContentAPI/modify.ts:39]: why to recreate the fold func each time? i'd move it out and call `foldTree`, like its couterpart `unfoldTree`

**Review by lxgreen — CHANGES_REQUESTED** [review]

File: packages/ricos-content/web/src/RicosContentAPI/modify.spec.ts [hunk 6]
```diff
@@ -1,20 +1,14 @@
-import { cloneDeep } from 'lodash';
 import { modify } from './modify';
+import { compare } from '..';
 import { Node_Type, RichContent } from 'ricos-schema';
-import rawContent from '../../statics/json/migratedFixtures/migration-content.json';
+import rawContent from '../../statics/json/modifyFixtures/images-dividers.json';
```
Comment (lxgreen) [inline packages/ricos-content/web/src/RicosContentAPI/modify.spec.ts:4]: everything in /statics is copied to /dist and bundled. it's better to move fixtures to /tests

File: packages/ricos-content/web/src/RicosContentAPI/modify.spec.ts [hunk 7]
```diff
@@ -27,25 +21,16 @@ describe('Content modify', () => {
...
-    });
+    const expected = RichContent.fromJSON(rawContentDoubleDividers);
 
     const actual = modify(content)
       .filter(({ type }) => type === Node_Type.DIVIDER)
       .set(n => [n, n]);
 
-    expect(actual).toStrictEqual(expected);
-    expect(actual).not.toEqual(content);
+    expect(compare(actual, expected, { ignoredKeys: ['key'] })).toEqual({});
```
Comment (lxgreen) [inline packages/ricos-content/web/src/RicosContentAPI/modify.spec.ts:34]: wuuuut? no no
i meant, the 
```js
expect(actual).toStrictEqual(expected);
```
is enough for test. The `compare` is buggy itself, and we are should try to decrease complexity in tests.

File: packages/ricos-content/web/src/RicosContentAPI/modify.ts [hunk 8]
```diff
@@ -3,16 +3,28 @@ import { compact, isArray } from 'lodash';
 import { Prism, fromTraversable, Traversal } from 'monocle-ts';
 import { RichContent, Node, Node_Type } from 'ricos-schema';
 
+type ModifierSetter = (node: Node) => Node | Node[];
 export interface Modifier {
   filter: (pred: (node: Node) => boolean) => Modifier;
-  set: (setter: (node: Node) => Node | Node[]) => RichContent;
+  set: (setter: ModifierSetter) => RichContent;
```
Comment (lxgreen) [inline packages/ricos-content/web/src/RicosContentAPI/modify.ts:9]: nah, dedicated type for `node => node` ? better no

File: packages/ricos-content/web/src/RicosContentAPI/modify.ts [hunk 9]
```diff
@@ -3,16 +3,28 @@ import { compact, isArray } from 'lodash';
...
+const foldTree = (tree: T.Tree<Node>, setter: ModifierSetter, keysToSet: string[]) =>
+  T.fold<Node, Node>((value, forest) => {
+    const nodes = forest.reduce((prev, curr) => {
+      const mapped = keysToSet.includes(curr.key) ? setter(curr) : curr;
+      const item = isArray(mapped) ? mapped : [mapped];
+      return [...prev, ...item];
+    }, [] as Node[]);
+    value.nodes = nodes;
+    return value;
+  })(tree);
```
Comment (lxgreen) [inline packages/ricos-content/web/src/RicosContentAPI/modify.ts:26]:
```suggestion
// import { curry } from 'lodash/fp';
// import { pipe } from 'fp-ts/lib/function';

const toArray = item => isArray(item) ? item : [item];

const modifyByKey = curry((keysToSet: string[], setter: (node: Node) => Node | Node[], node: Node) =>
  keysToSet.includes(node.key) ? setter(node) : node);

const mergeWith = curry((prefix, suffix) => [...prefix, ...suffix]);

const foldTree = (tree: T.Tree<Node>, setter: ModifierSetter, keysToSet: string[]) =>
  T.fold<Node, Node>((root, forest) => ({
    ...root,
    nodes: forest.reduce((modifiedForest, node) =>
      pipe(node, modifyByKey(keysToSet, setter), toArray, mergeWith(modifiedForest)), [] as Node[])
  }))(tree);

```

Comment (lxgreen) [reply packages/ricos-content/web/src/RicosContentAPI/modify.ts:26]: a bit more fun

**Review by lxgreen — CHANGES_REQUESTED** [review]
last cycle, promise!

File: packages/ricos-content/web/src/RicosContentAPI/modify.ts (hunk 9 above)
Comment (Barackos) [reply packages/ricos-content/web/src/RicosContentAPI/modify.ts]: Wow! 🎉

**Pull request merged**